Show a clear error if the key is found but the slots are not configured

w15eacre opened the pull request:
- Fix the error message when a key is connected, but no slots are configured
- The solution is to check connected devices. If they exist, show the error: "Hardware keys found, but no slots are configured."
- Fix missing SCardDisconnect
- Fix outdated test cases

Fixes issue #11543


## Screenshots
![image](https://github.com/user-attachments/assets/ce3e1f08-db44-4979-88ea-78a14a42bd2a)

![image](https://github.com/user-attachments/assets/aa1354ca-a722-4800-88e6-08ce0a2c23ac)


## Testing strategy
- Check the fix on YubiKeys
- Add a unit test to verify the series numbers of connected devices


## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)

Commit:
Show a clear error if the key is found but the slots are not configured

- Fix missed SCardDisconnect
- Fix outdated test

Fixes #11543

Commit:
Fix build in ubuntu

Commit:
Fix static_assert message

droidmonkey commented:
Our minimum version of Qt is 5.12 now so you should be able to use `qScopeGuard` template function instead of redefining a `ScopeGuard`. https://doc.qt.io/qt-5/qscopeguard.html#qScopeGuardx

This message should also appear on the database open widget. We can show the same disabled drop down box as the database security view and also disable the checkbox too.

w15eacre commented:
I tried the first time but Ubuntu build failed on CI

droidmonkey commented:
Note the case, qScopeGuard not QScopeGuard

Commit:
Remove the ScopedGuard and replace it with the qScopeGuard function.

Fixes #11543

Commit:
Add the error message about the slot on the Open Database page.

Fixes #11543

w15eacre commented:
@droidmonkey Thanks for your review; I have fixed it.
In my opinion, If I added a new item in the combobox I would refactor the noHardwareKeysFoundLabel and move it into this combobox. This behavior would be the same as in the Database Security view.

droidmonkey commented:
I don't like that the keys are basically polled twice just to get an error message. Instead, I tweaked the `findValidKeys` function to include a reference parameter that populates the number of keys actually found. This is then stored in the Yubikey class for use in the new `connectedKeys` function. This actually unwinds a lot of the new code you provided, but is way more elegant.

What do you think?

droidmonkey committed:
Avoid double polling hardware keys

* Also fix delayed polling on window activation

w15eacre commented:
Yes, it's better and safer. Besides, polling twice caused an issue with the calling function that could freeze the UI thread. I suggest replacing:
- int& → unsigned int8_t& or size_t&. This value is in the range [0, 4).
- int Delay → std::chrono::milliseconds for a clearer understanding of this interval.

@droidmonkey What do you think about it?

droidmonkey commented:
> int& → unsigned int8_t& or size_t&. This value is in the range [0, 4).

The sign of the counter variable is irrelevant in this context and can actually produce underflow errors if you were to code in a -- by accident (ie, instead of -1 you'll get UINT_MAX). Either way this won't happen or matter because that isn't how this variable is handled in any context of the code. Generally speaking, the "only" time you need to use unsigned int is to use the full 32-bit length of an int to represent the numbers you need.

> int Delay → std::chrono::milliseconds for a clearer understanding of this interval.

You cannot do this because std::chrono::milliseconds is actually a class and not convertible to an int to be used with Qt. It is not meant to indicate the variable contents by a name.

droidmonkey commented:
Also, I tested this code with my 5 yubikeys and NFC reader, it worked flawlessly and reliably. Merging now.
